Flaky test_participant_selection

```
______________ test_participant_selection[matrix-False-3-10-0-6] _______________

raiden_apps = [<App 0x68299b638f40284f4d0F3ed99fF8aa6fA1FB69D4>, <App 0x54eFD192F7c8Eb5Eb3E2a365a9fcC099661eb3DE>, <App 0xf4D00AbCa9...7768fF107250c5FD7>, <App 0xE3da17c306548aE93E867E0156A98996715d50a7>, <App 0xA0cD5e31Ff9665F9cA128f7854D069A27c10DFC2>]
test_function = <function run_test_participant_selection at 0x7f03fb6e2d90>
kwargs = {'raiden_network': [<App 0x68299b638f40284f4d0F3ed99fF8aa6fA1FB69D4>, <App 0x54eFD192F7c8Eb5Eb3E2a365a9fcC099661eb3DE>...65F9cA128f7854D069A27c10DFC2>], 'token_addresses': [b'\n\x1d\x9e\r\xd2\x8e\x7f\x0c\t\xe0\xf9h\x17\xbe\x88\x8fn\x0bL*']}
result = <gevent._event.AsyncResult object at 0x7f03f9826048>
app = <App 0xA0cD5e31Ff9665F9cA128f7854D069A27c10DFC2>
test_greenlet = <Greenlet "Greenlet-3" at 0x7f03f93606a8: run_test_participant_selection(raiden_network=[<App 0x68299b638f40284f4d0F3ed99fF8aa6fA1FB69D4>,, token_addresses=[b'\n\x1d\x9e\r\xd2\x8e\x7f\x0c\t\xe0\xf9h\x17\xbe)>

    def raise_on_failure(raiden_apps, test_function, **kwargs):
        """Wait on the result for the test function and any of the apps.
    
        This utility should be used for happy path testing with more than one app.
        This will raise if any of the apps is killed.
        """
        result = AsyncResult()
    
        # Do not use `link` or `link_value`, an app an be stopped to test restarts.
        for app in raiden_apps:
            assert app.raiden, "The RaidenService must be started"
            app.raiden.link_exception(result)
    
        test_greenlet = gevent.spawn(test_function, **kwargs)
        test_greenlet.link(result)
    
        # Returns if either happens:
        # - The test finished (successfully or not)
        # - One of the apps crashed during the test
        try:
>           result.get()

raiden/tests/utils/detect_failure.py:30: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/event.py:381: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/event.py:409: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/event.py:399: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/event.py:379: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

t = <class 'raiden.exceptions.RaidenRecoverableError'>
value = RaidenRecoverableError('Creating a new channel will fail - Gas estimation failed for unknown reason.')
tb = <traceback object at 0x7f03f9149a08>

    def reraise(t, value, tb=None): # pylint:disable=unused-argument
        if value.__traceback__ is not tb and tb is not None:
>           raise value.with_traceback(tb)

../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/_compat.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/greenlet.py:716: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   self.stop_event.wait()

raiden/raiden_service.py:452: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/event.py:240: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/event.py:140: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/event.py:117: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/event.py:119: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/_greenlet_primitives.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/_greenlet_primitives.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???

src/gevent/_greenlet_primitives.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   raiden.exceptions.RaidenRecoverableError: Creating a new channel will fail - Gas estimation failed for unknown reason.

src/gevent/__greenlet_primitives.pxd:35: RaidenRecoverableError

During handling of the above exception, another exception occurred:

raiden_network = [<App 0x68299b638f40284f4d0F3ed99fF8aa6fA1FB69D4>, <App 0x54eFD192F7c8Eb5Eb3E2a365a9fcC099661eb3DE>, <App 0xf4D00AbCa9...7768fF107250c5FD7>, <App 0xE3da17c306548aE93E867E0156A98996715d50a7>, <App 0xA0cD5e31Ff9665F9cA128f7854D069A27c10DFC2>]
token_addresses = [b'\n\x1d\x9e\r\xd2\x8e\x7f\x0c\t\xe0\xf9h\x17\xbe\x88\x8fn\x0bL*']

    @pytest.mark.parametrize("number_of_nodes", [6])
    @pytest.mark.parametrize("channels_per_node", [0])
    @pytest.mark.parametrize("settle_timeout", [10])
    @pytest.mark.parametrize("reveal_timeout", [3])
    def test_participant_selection(raiden_network, token_addresses):
        raise_on_failure(
            raiden_network,
            run_test_participant_selection,
            raiden_network=raiden_network,
>           token_addresses=token_addresses,
        )

raiden/tests/integration/long_running/test_token_networks.py:78: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
raiden/tests/utils/detect_failure.py:41: in raise_on_failure
    log.exception("Pending greenlets", tracebacks="\n".join(gevent.util.format_run_info()))
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:154: in format_run_info
    _format_greenlet_info(lines, greenlet_stacks, limit)
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:198: in _format_greenlet_info
    'running_stack_limit': limit,
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:343: in format_lines
    for l in self._render(tree)]
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:415: in _render
    self.__render_children(tree)
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:427: in __render_children
    child_tree = child._render(tree.deeper())
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:415: in _render
    self.__render_children(tree)
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:427: in __render_children
    child_tree = child._render(tree.deeper())
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:415: in _render
    self.__render_children(tree)
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:427: in __render_children
    child_tree = child._render(tree.deeper())
../venv-3.7-LINUX/lib/python3.7/site-packages/gevent/util.py:380: in _render
    label = repr(self.greenlet)
src/gevent/greenlet.py:430: in gevent._greenlet.Greenlet.__repr__
    ???
src/gevent/greenlet.py:436: in gevent._greenlet.Greenlet._formatinfo
    ???
src/gevent/greenlet.py:436: in gevent._greenlet.Greenlet._formatinfo
    ???
src/gevent/greenlet.py:448: in gevent._greenlet.Greenlet._formatinfo
    ???
raiden/connection_manager.py:462: in __repr__
    chain_state=views.state_from_raiden(self.raiden),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

raiden = <RaidenService node:0xA0cD5e31Ff9665F9cA128f7854D069A27c10DFC2>

    def state_from_raiden(raiden) -> ChainState:  # pragma: no unittest
>       return raiden.wal.state_manager.current_state
E       AttributeError: 'NoneType' object has no attribute 'state_manager'

raiden/transfer/views.py:81: AttributeError
```

Seen in https://circleci.com/gh/raiden-network/raiden/62171#tests/containers/15

[raiden-debug_2019-07-05T07:42:05.086997.log](https://github.com/raiden-network/raiden/files/3361929/raiden-debug_2019-07-05T07.42.05.086997.log)
[results.zip](https://github.com/raiden-network/raiden/files/3361934/results.zip)

--- a/raiden/api/python.py
+++ b/raiden/api/python.py
@@ -322,6 +322,22 @@ class RaidenAPI:  # pragma: no unittest
 
         return connection_manager.leave(registry_address)
 
+    def is_already_existing_channel(
+        self,
+        token_network_address: TokenNetworkAddress,
+        partner_address: Address,
+        block_identifier: Optional[BlockSpecification] = None,
+    ) -> bool:
+        chain_state = self.raiden.chain
+        proxy = chain_state.address_to_token_network[token_network_address]
+        channel_identifier = proxy.get_channel_identifier_or_none(
+            participant1=self.raiden.address,
+            participant2=partner_address,
+            block_identifier=block_identifier or chain_state.client.get_checking_block(),
+        )
+
+        return channel_identifier is not None
+
     def channel_open(
         self,
         registry_address: PaymentNetworkAddress,
@@ -358,18 +374,18 @@ class RaidenAPI:  # pragma: no unittest
                 "Token network for token %s does not exist" % to_checksum_address(token_address)
             )
 
-        token_network = self.raiden.chain.token_network(registry.get_token_network(token_address))
+        token_network = self.raiden.chain.token_network(token_network_address)
+        given_block_identifier = views.state_from_raiden(self.raiden).block_hash
 
-        token_network_proxy = self.raiden.chain.address_to_token_network[token_network_address]
-        channel_identifier = token_network_proxy.get_channel_identifier_or_none(
-            participant1=self.raiden.address,
-            participant2=partner_address,
-            block_identifier=token_network.client.get_checking_block(),
+        duplicated_channel = self.is_already_existing_channel(
+            token_network_address=token_network_address,
+            partner_address=partner_address,
+            block_identifier=given_block_identifier,
         )
-
-        if channel_identifier:
-            raise RecoverableDuplicatedChannelError(
-                "Channel with given partner address already exists"
+        if duplicated_channel:
+            raise DuplicatedChannelError(
+                f"A channel with {partner_address} for token {token_address} already exists. "
+                f"(At block: {given_block_identifier})"
             )
 
         with self.raiden.gas_reserve_lock:
@@ -388,10 +404,19 @@ class RaidenAPI:  # pragma: no unittest
                 token_network.new_netting_channel(
                     partner=partner_address,
                     settle_timeout=settle_timeout,
-                    given_block_identifier=views.state_from_raiden(self.raiden).block_hash,
+                    given_block_identifier=given_block_identifier,
                 )
             except DuplicatedChannelError:
                 log.info("partner opened channel first")
+            except RaidenRecoverableError:
+                # The channel may have been created in the pending block.
+                duplicated_channel = self.is_already_existing_channel(
+                    token_network_address=token_network_address, partner_address=partner_address
+                )
+                if duplicated_channel:
+                    log.info("partner opened channel first")
+                else:
+                    raise
 
         waiting.wait_for_newchannel(
             raiden=self.raiden,

--- a/raiden/api/python.py
+++ b/raiden/api/python.py
@@ -350,17 +350,6 @@ class RaidenAPI:  # pragma: no unittest
         if not is_binary_address(partner_address):
             raise InvalidAddress("Expected binary address format for partner in channel open")
 
-        chain_state = views.state_from_raiden(self.raiden)
-        channel_state = views.get_channelstate_for(
-            chain_state=chain_state,
-            payment_network_address=registry_address,
-            token_address=token_address,
-            partner_address=partner_address,
-        )
-
-        if channel_state:
-            raise DuplicatedChannelError("Channel with given partner address already exists")
-
         registry = self.raiden.chain.token_network_registry(registry_address)
         token_network_address = registry.get_token_network(token_address)
 
@@ -370,6 +359,18 @@ class RaidenAPI:  # pragma: no unittest
             )
 
         token_network = self.raiden.chain.token_network(registry.get_token_network(token_address))
+
+        token_network_proxy = self.raiden.chain.address_to_token_network[token_network_address]
+        channel_identifier = token_network_proxy.get_channel_identifier_or_none(
+            participant1=self.raiden.address,
+            participant2=partner_address,
+            block_identifier=token_network.client.get_checking_block(),
+        )
+
+        if channel_identifier:
+            raise RecoverableDuplicatedChannelError(
+                "Channel with given partner address already exists"
+            )
 
         with self.raiden.gas_reserve_lock:
             has_enough_reserve, estimated_required_reserve = has_enough_gas_reserve(

--- a/raiden/connection_manager.py
+++ b/raiden/connection_manager.py
@@ -22,7 +22,7 @@ from raiden.exceptions import (
 from raiden.settings import DEFAULT_RETRY_TIMEOUT
 from raiden.transfer import views
 from raiden.utils import typing
-from raiden.utils.typing import Address, BlockNumber, TokenAmount, TokenNetworkAddress
+from raiden.utils.typing import Address, TokenAmount, TokenNetworkAddress
 
 log = structlog.get_logger(__name__)
 RECOVERABLE_ERRORS = (
@@ -326,17 +326,6 @@ class ConnectionManager:  # pragma: no unittest
             )
             if not channel_identifier:
                 raise
-
-        # Wait until a newer block is mined before moving to deposit.
-        # TODO: Remove this waiting block when
-        # https://github.com/raiden-network/raiden/issues/4220
-        # is resolved.
-        chain_state = views.state_from_raiden(self.raiden)
-        waiting.wait_for_block(
-            raiden=self.raiden,
-            block_number=BlockNumber(chain_state.block_number + 1),
-            retry_timeout=DEFAULT_RETRY_TIMEOUT,
-        )
 
         total_deposit = self._initial_funding_per_partner
         if total_deposit == 0:

--- a/raiden/connection_manager.py
+++ b/raiden/connection_manager.py
@@ -19,7 +19,6 @@ from raiden.exceptions import (
     RaidenUnrecoverableError,
     TransactionThrew,
 )
-from raiden.settings import DEFAULT_RETRY_TIMEOUT
 from raiden.transfer import views
 from raiden.utils import typing
 from raiden.utils.typing import Address, TokenAmount, TokenNetworkAddress
@@ -310,36 +309,10 @@ class ConnectionManager:  # pragma: no unittest
             # If channel already exists (either because partner created it,
             # or it's nonfunded channel), continue to ensure it's funded
             pass
-        except RaidenRecoverableError:
-            # The channel may still have been created in the pending block.
-            # If that is the case, swallow the exception
-            token_network_address = views.get_token_network_address_by_token_address(
-                chain_state=views.state_from_raiden(self.raiden),
-                payment_network_address=self.registry_address,
-                token_address=self.token_address,
-            )
-            token_network = self.raiden.chain.address_to_token_network[token_network_address]
-            channel_identifier = token_network.get_channel_identifier_or_none(
-                participant1=self.raiden.address,
-                participant2=partner,
-                block_identifier=token_network.client.get_checking_block(),
-            )
-            if not channel_identifier:
-                raise
 
         total_deposit = self._initial_funding_per_partner
         if total_deposit == 0:
             return
-
-        # The channel may still be in the pending block only and our node may
-        # not be aware of it.
-        waiting.wait_for_newchannel(
-            raiden=self.raiden,
-            payment_network_address=self.registry_address,
-            token_address=self.token_address,
-            partner_address=partner,
-            retry_timeout=DEFAULT_RETRY_TIMEOUT,
-        )
 
         try:
             self.api.set_total_channel_deposit(

--- a/raiden/connection_manager.py
+++ b/raiden/connection_manager.py
@@ -310,6 +310,22 @@ class ConnectionManager:  # pragma: no unittest
             # If channel already exists (either because partner created it,
             # or it's nonfunded channel), continue to ensure it's funded
             pass
+        except RaidenRecoverableError:
+            # The channel may still have been created in the pending block.
+            # If that is the case, swallow the exception
+            token_network_address = views.get_token_network_address_by_token_address(
+                chain_state=views.state_from_raiden(self.raiden),
+                payment_network_address=self.registry_address,
+                token_address=self.token_address,
+            )
+            token_network = self.raiden.chain.address_to_token_network[token_network_address]
+            channel_identifier = token_network.get_channel_identifier_or_none(
+                participant1=self.raiden.address,
+                participant2=partner,
+                block_identifier=token_network.client.get_checking_block(),
+            )
+            if not channel_identifier:
+                raise
 
         # Wait until a newer block is mined before moving to deposit.
         # TODO: Remove this waiting block when
@@ -325,6 +341,16 @@ class ConnectionManager:  # pragma: no unittest
         total_deposit = self._initial_funding_per_partner
         if total_deposit == 0:
             return
+
+        # The channel may still be in the pending block only and our node may
+        # not be aware of it.
+        waiting.wait_for_newchannel(
+            raiden=self.raiden,
+            payment_network_address=self.registry_address,
+            token_address=self.token_address,
+            partner_address=partner,
+            retry_timeout=DEFAULT_RETRY_TIMEOUT,
+        )
 
         try:
             self.api.set_total_channel_deposit(

--- a/raiden/connection_manager.py
+++ b/raiden/connection_manager.py
@@ -458,6 +458,11 @@ class ConnectionManager:  # pragma: no unittest
         return self.initial_channel_target < 1
 
     def __repr__(self) -> str:
+        if self.raiden.wal is None:
+            return (
+                f"{self.__class__.__name__}(target={self.initial_channel_target} "
+                "WAL not initialized)"
+            )
         open_channels = views.get_channelstate_open(
             chain_state=views.state_from_raiden(self.raiden),
             payment_network_address=self.registry_address,

--- a/raiden/tests/integration/long_running/test_token_networks.py
+++ b/raiden/tests/integration/long_running/test_token_networks.py
@@ -66,7 +66,6 @@ def saturated_count(connection_managers, registry_address, token_address):
 # - Check if this test needs to be adapted for the matrix transport
 #   layer when activating it again. It might as it depends on the
 #   raiden_network fixture.
-@pytest.mark.flaky(max_runs=5)
 @pytest.mark.parametrize("number_of_nodes", [6])
 @pytest.mark.parametrize("channels_per_node", [0])
 @pytest.mark.parametrize("settle_timeout", [10])

--- a/raiden/tests/integration/long_running/test_token_networks.py
+++ b/raiden/tests/integration/long_running/test_token_networks.py
@@ -63,9 +63,6 @@ def saturated_count(connection_managers, registry_address, token_address):
 # TODO: add test scenarios for
 # - subsequent `connect()` calls with different `funds` arguments
 # - `connect()` calls with preexisting channels
-# - Check if this test needs to be adapted for the matrix transport
-#   layer when activating it again. It might as it depends on the
-#   raiden_network fixture.
 @pytest.mark.parametrize("number_of_nodes", [6])
 @pytest.mark.parametrize("channels_per_node", [0])
 @pytest.mark.parametrize("settle_timeout", [10])

